### PR TITLE
grpc auth: add file based grpc credentials extension

### LINKED ISSUE #3392: GrpcService CallCredentials loads credentials from filesystem
**Issue Template**

*Title:  Add a `FileBasedMetadataCredentialsPlugin` to load credential from a file as CallCredentials*

*Description*:

**What we want to do**
Envoy recently adds the support,  to implement customized CallCredentials as MetadataCredentialsPlugin, https://github.com/envoyproxy/envoy/pull/3161.

We'd like to add an implementation of `FileBasedMetadataCredentialsPlugin` (just an illustrative name)which can:

1. Configured with a path to the credentials file.
1. When `FileBasedMetadataCredentialPlugin::GetMetadata` is invoked, the plugin reads the credential file, and populates the metadata with the file's content.
1. The outcome is that every xDS request includes this credential as metadata header.

**Use Case**
For our use case,
1. We run envoy on Kubernetes. And Kubernetes mounts [a credential file](https://github.com/mikedanese/community/blob/6e209490c441d8df84b6b5d8e352c0e2491a41bd/contributors/design-proposals/auth/bound-service-account-tokens.md)(in format of a JWT) into the envoy container file system.
1. Envoy sends SDS request to an identity server to get Istio key and certificate.
1. The SDS service will be configured with `FileBasedMetadataCredentialsPlugin` proposed above, and load the credential from the file, put in the metadata header.
1. The identity server then can verify the credential and authenticate the envoy, issue the key cert.

That describes Istio workload identity provisioning process. However,

- The JWT credential is provided by native Kubernetes, anyone who wants  make Envoy xDS requests including this Kubernetes-credentials can benefit from such feature.
- Potentially, it can be used beyong Kubernetes as long as a platform can provide a bootstrap credential as a file.

[optional *Relevant Links*:]

1. [Kubernetes JWT design doc](https://github.com/mikedanese/community/blob/6e209490c441d8df84b6b5d8e352c0e2491a41bd/contributors/design-proposals/auth/bound-service-account-tokens.md)

### PR DESCRIPTION
**Pull request by wozz**
*Description*:
Adds a file based grpc credentials extension.  See issue #3392 for more details.

*Risk Level*: Low: extension for grpc credentials loaded by explicit configuration options

*Testing*: tests included in PR

*Docs Changes*: Inline docs via comments and proto docs

*Release Notes*: N/A

Fixes #3392

### CODE REVIEW

**Comment by htuch** [comment]
@lizan can you verify this meets the needs we have in #3392?

File: test/common/grpc/grpc_client_integration_test.cc [hunk 1]
```diff
@@ -949,6 +951,125 @@ TEST_P(GrpcAccessTokenClientIntegrationTest, InvalidCredentialFactory) {
   EXPECT_THROW_WITH_MESSAGE(initialize(), EnvoyException,
                             "Unknown google grpc credentials factory: unknown");
 }
+
```
Comment (lizan) [inline test/common/grpc/grpc_client_integration_test.cc]: Move these tests into a new target under extensions?

File: source/extensions/grpc_credentials/file_based_metadata/config.cc [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+FileBasedMetadataAuthenticator::GetMetadata(grpc::string_ref, grpc::string_ref,
+                                            const grpc::AuthContext&,
+                                            std::multimap<grpc::string, grpc::string>* metadata) {
+  std::string header_value = Envoy::Config::DataSource::read(config_.secret_data(), true);
+  std::string header_key = "authorization";
+  std::string header_prefix = "";
+  if (!config_.header_key().empty()) {
+    header_key = config_.header_key();
+  }
+  if (!config_.header_prefix().empty()) {
```
Comment (lizan) [inline source/extensions/grpc_credentials/file_based_metadata/config.cc]: no need if, you can just do `std::string header_prefix = config_.header_prefix()`

File: api/envoy/extensions/grpc_credentials/file_based_metadata.proto [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  // this secret will be attached to a header of the gRPC connection
+  envoy.api.v2.core.DataSource secret_data = 1;
+
+  // Metadata header key to use for sending the secret data
+  // if no header key is set, "authorization" header will be used
+  string header_key = 2;
+
+  // Prefix to prepend to the secret in the metadata header
+  // if no prefix is set, the default is to use no prefix
+  string header_prefix = 3;
```
Comment (htuch) [inline api/envoy/extensions/grpc_credentials/file_based_metadata.proto]: Is this (prefix) going to be flexible enough for anticipated future needs?

File: source/extensions/grpc_credentials/file_based_metadata/config.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ * This implementation uses ssl creds for the grpc channel if available. Additionally, it uses
+ * MetadataCredentialsFromPlugin to add a static secret that is loaded from a file. The header key
+ * and header prefix are configurable.
+ *
+ * This implementation uses the from_plugin field in the call credentials config to get the filename
+ * of where the secret is stored to add to the header.
+ */
+class FileBasedMetadataGrpcCredentialsFactory : public Grpc::GoogleGrpcCredentialsFactory {
+public:
+  virtual std::shared_ptr<grpc::ChannelCredentials>
```
Comment (htuch) [inline source/extensions/grpc_credentials/file_based_metadata/config.h]: You can elide the `virtual` if you have `overrride`.

File: test/extensions/grpc_credentials/file_based_metadata/file_based_metadata_grpc_credentials_test.cc [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+  virtual envoy::api::v2::core::GrpcService createGoogleGrpcConfig() override {
+    auto config = GrpcClientIntegrationTest::createGoogleGrpcConfig();
+    auto* google_grpc = config.mutable_google_grpc();
+    google_grpc->set_credentials_factory_name(credentials_factory_name_);
+    auto* ssl_creds = google_grpc->mutable_channel_credentials()->mutable_ssl_credentials();
+    ssl_creds->mutable_root_certs()->set_filename(
+        TestEnvironment::runfilesPath("test/config/integration/certs/upstreamcacert.pem"));
+    if (header_value_1_ != "") {
+      std::string yaml1 = fmt::sprintf(R"EOF(
```
Comment (htuch) [inline test/extensions/grpc_credentials/file_based_metadata/file_based_metadata_grpc_credentials_test.cc]: FWIW, most tests use `fmt::format` for this kind of thing.

File: test/extensions/grpc_credentials/file_based_metadata/file_based_metadata_grpc_credentials_test.cc [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+header_key: %s
+header_prefix: %s
+)EOF",
+                                       header_value_1_, header_key_1_, header_prefix_1_);
+      auto* plugin_config = google_grpc->add_call_credentials()->mutable_from_plugin();
+      plugin_config->set_name(credentials_factory_name_);
+      envoy::extensions::grpc_credentials::FileBasedMetadataConfig metadata_config;
+      MessageUtil::loadFromYaml(yaml1, *plugin_config->mutable_config());
+    }
+    if (header_value_2_ != "") {
```
Comment (htuch) [inline test/extensions/grpc_credentials/file_based_metadata/file_based_metadata_grpc_credentials_test.cc]: `!header_value_2.empty()` (and elsewhere)

File: test/extensions/grpc_credentials/file_based_metadata/file_based_metadata_grpc_credentials_test.cc (hunk 5 above)
Comment (htuch) [inline test/extensions/grpc_credentials/file_based_metadata/file_based_metadata_grpc_credentials_test.cc]: Nit: const

File: source/extensions/grpc_credentials/file_based_metadata/config.h [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  getChannelCredentials(const envoy::api::v2::core::GrpcService& grpc_service_config) override;
+
+  Envoy::ProtobufTypes::MessagePtr createEmptyConfigProto() {
+    return Envoy::ProtobufTypes::MessagePtr{
```
Comment (htuch) [inline source/extensions/grpc_credentials/file_based_metadata/config.h]: Nit: `std::make_unique`

File: source/extensions/grpc_credentials/file_based_metadata/config.h [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace Envoy {
+namespace Extensions {
+namespace GrpcCredentials {
+namespace FileBasedMetadata {
+
+/**
+ * File Based Metadata implementation of Google Grpc Credentials Factory
+ * This implementation uses ssl creds for the grpc channel if available. Additionally, it uses
+ * MetadataCredentialsFromPlugin to add a static secret that is loaded from a file. The header key
```
Comment (htuch) [inline source/extensions/grpc_credentials/file_based_metadata/config.h]: I wonder if you want to be able to compose these together. It seems kind of redundant to be repeating the SSL creds optional logic in every gRPC creds extensions, and maybe you want a file metadata credential factory that can be combined with some other channel creds (e.g. ALTS)? Any thoughts on how to make these factories more composable?

File: api/envoy/extensions/grpc_credentials/file_based_metadata.proto [hunk 9]
```diff
@@ -0,0 +1,26 @@
+syntax = "proto3";
+
+// [#protodoc-title: Grpc Credentials File Based Metadata]
+// Configuration for File Based Metadata Grpc Credentials Plugin
+
+package envoy.extensions.grpc_credentials;
```
Comment (htuch) [inline api/envoy/extensions/grpc_credentials/file_based_metadata.proto]: Also, this should live in the https://github.com/envoyproxy/envoy/tree/master/api/envoy/config subtree.

File: api/envoy/extensions/grpc_credentials/file_based_metadata.proto (hunk 3 above)
Comment (wozz) [reply api/envoy/extensions/grpc_credentials/file_based_metadata.proto]: I included this mostly to support the `Authorization: Bearer <token>` style headers. Are there other common patterns that need more configurable options to be supported?

Comment (htuch) [reply api/envoy/extensions/grpc_credentials/file_based_metadata.proto]: IDK; @lizan @incfly ?

**Pull request merged**